### ISSUE TITLE
ROW_NUMBER OVER PARTITION statements being generated for MySQL < 8.0

### ISSUE THREAD

**Issue by fmauNeko** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#864]
### Steps to reproduce
Use MySQL < 8.0 (5.7.27 in that case)
Execute following query
```c#
var resaLocat = readContext.Resa
  .FromSqlInterpolated($@"
  (
    SELECT r.*
    FROM resa r JOIN lots l ON l.RefAgence = r.RefAgence AND l.No_Lot = r.No_Dossier_Lot
    WHERE
      r.RefAgence = {agencyId}
      AND r.Etat_Resa BETWEEN 2 AND 4
      AND STR_TO_DATE(r.Date_Debut, '%Y/%m/%d') <= {occ.Date}
      AND STR_TO_DATE(r.Date_Fin, '%Y/%m/%d') > {occ.Date}
      AND l.Ville = {occ.Station}
      AND l.Situation = {occ.Quartier}
  )
  UNION
  (
    SELECT r.*
    FROM resa r JOIN lots_nonweb l ON l.RefAgence = r.RefAgence AND l.No_Lot = r.No_Dossier_Lot
    WHERE
      r.RefAgence = {agencyId}
      AND r.Etat_Resa BETWEEN 2 AND 4
      AND STR_TO_DATE(r.Date_Debut, '%Y/%m/%d') <= {occ.Date}
      AND STR_TO_DATE(r.Date_Fin, '%Y/%m/%d') > {occ.Date}
      AND l.Ville = {occ.Station}
      AND l.Situation = {occ.Quartier}
  )
  ")
  .Select(resa => new
  {
    resa,
    locat = readContext.Locat
      .SingleOrDefault(l => l.RefAgence == resa.RefAgence && l.NoLocataire == resa.NoLocataire),
  })
  .ToList();
```

### The issue
The generated query includes a ```ROW_NUMBER() OVER(PARTITION BY `l`.`RefAgence`, `l`.`No_Locataire` ORDER BY `l`.`RefAgence`, `l`.`No_Locataire`)``` statement, which isn't supported by MySQL < 8.0.

Executing the `ToList` before the `Select` solves the issue.

```
fail: Microsoft.EntityFrameworkCore.Database.Command[20102]
      Failed executing DbCommand (26ms) [Parameters=[p0='?' (Size = 4000), p1='?' (DbType = DateTime), p2='?' (DbType = DateTime), p3='?' (Size = 4000), p4='?' (Size = 4000), p5='?' (Size = 4000), p6='?' (DbType = DateTime), p7='?' (DbType = DateTime), p8='?' (Size = 4000), p9='?' (Size = 4000)], CommandType='Text', CommandTimeout='30']
      SELECT `r`.`RefAgence`, `r`.`No_Resa`, `r`.`BRUT_PROPRIETAIRE`, `r`.`COUT_ASSUR_TTC`, `r`.`COUT_FDD_TTC`, `r`.`COUT_HEBERGEMENT`, `r`.`COUT_MENAGE_TTC`, `r`.`COUT_OPTIONS_TTC`, `r`.`Date_Debut`, `r`.`Date_Fin`, `r`.`Date_Option`, `r`.`Date_Resa`, `r`.`Etat_Resa`, `r`.`MONTANT1`, `r`.`MONTANT10`, `r`.`MONTANT11`, `r`.`MONTANT12`, `r`.`MONTANT13`, `r`.`MONTANT14`, `r`.`MONTANT15`, `r`.`MONTANT16`, `r`.`MONTANT17`, `r`.`MONTANT18`, `r`.`MONTANT19`, `r`.`MONTANT2`, `r`.`MONTANT20`, `r`.`MONTANT21`, `r`.`MONTANT22`, `r`.`MONTANT23`, `r`.`MONTANT24`, `r`.`MONTANT25`, `r`.`MONTANT26`, `r`.`MONTANT27`, `r`.`MONTANT3`, `r`.`MONTANT4`, `r`.`MONTANT5`, `r`.`MONTANT6`, `r`.`MONTANT7`, `r`.`MONTANT8`, `r`.`MONTANT9`, `r`.`Montant_Arrondi`, `r`.`MONTANT_REGLE`, `r`.`Nego`, `r`.`Net_Arrondi`, `r`.`NET_LOCATAIRE`, `r`.`NET_PROPRIETAIRE`, `r`.`NO_CE`, `r`.`No_Dossier_Lot`, `r`.`No_Locataire`, `r`.`NomLocataire`, `r`.`TAXE_SEJOUR`, `r`.`Type_Cogestion`, `t0`.`RefAgence`, `t0`.`No_Locataire`, `t0`.`Adresse`, `t0`.`Animaux`, `t0`.`AnneeDerResa`, `t0`.`Civilite`, `t0`.`Classification`, `t0`.`CP`, `t0`.`Date_Dossier`, `t0`.`Etat_Avancement`, `t0`.`Fax`, `t0`.`Langue`, `t0`.`Mail`, `t0`.`Nationalite`, `t0`.`Nom`, `t0`.`Occupants`, `t0`.`Origine`, `t0`.`Pays`, `t0`.`Portable`, `t0`.`Pref_Fax`, `t0`.`Pref_Portable`, `t0`.`Pref_Tel_Bur`, `t0`.`Pref_Tel_Dom`, `t0`.`Prenom`, `t0`.`Ref_Locataire`, `t0`.`Suite_Adresse`, `t0`.`Tel_Bur`, `t0`.`Tel_Dom`, `t0`.`Ville`
      FROM (
      
                    (
                      SELECT r.*
                      FROM resa r JOIN lots l ON l.RefAgence = r.RefAgence AND l.No_Lot = r.No_Dossier_Lot
                      WHERE
                        r.RefAgence = @p0
                        AND r.Etat_Resa BETWEEN 2 AND 4
                        AND STR_TO_DATE(r.Date_Debut, '%Y/%m/%d') <= @p1
                        AND STR_TO_DATE(r.Date_Fin, '%Y/%m/%d') > @p2
                        AND l.Ville = @p3
                        AND l.Situation = @p4
                    )
                    UNION
                    (
                      SELECT r.*
                      FROM resa r JOIN lots_nonweb l ON l.RefAgence = r.RefAgence AND l.No_Lot = r.No_Dossier_Lot
                      WHERE
                        r.RefAgence = @p5
                        AND r.Etat_Resa BETWEEN 2 AND 4
                        AND STR_TO_DATE(r.Date_Debut, '%Y/%m/%d') <= @p6
                        AND STR_TO_DATE(r.Date_Fin, '%Y/%m/%d') > @p7
                        AND l.Ville = @p8
                        AND l.Situation = @p9
                    )
                    
      ) AS `r`
      LEFT JOIN (
          SELECT `t`.`RefAgence`, `t`.`No_Locataire`, `t`.`Adresse`, `t`.`Animaux`, `t`.`AnneeDerResa`, `t`.`Civilite`, `t`.`Classification`, `t`.`CP`, `t`.`Date_Dossier`, `t`.`Etat_Avancement`, `t`.`Fax`, `t`.`Langue`, `t`.`Mail`, `t`.`Nationalite`, `t`.`Nom`, `t`.`Occupants`, `t`.`Origine`, `t`.`Pays`, `t`.`Portable`, `t`.`Pref_Fax`, `t`.`Pref_Portable`, `t`.`Pref_Tel_Bur`, `t`.`Pref_Tel_Dom`, `t`.`Prenom`, `t`.`Ref_Locataire`, `t`.`Suite_Adresse`, `t`.`Tel_Bur`, `t`.`Tel_Dom`, `t`.`Ville`
          FROM (
              SELECT `l`.`RefAgence`, `l`.`No_Locataire`, `l`.`Adresse`, `l`.`Animaux`, `l`.`AnneeDerResa`, `l`.`Civilite`, `l`.`Classification`, `l`.`CP`, `l`.`Date_Dossier`, `l`.`Etat_Avancement`, `l`.`Fax`, `l`.`Langue`, `l`.`Mail`, `l`.`Nationalite`, `l`.`Nom`, `l`.`Occupants`, `l`.`Origine`, `l`.`Pays`, `l`.`Portable`, `l`.`Pref_Fax`, `l`.`Pref_Portable`, `l`.`Pref_Tel_Bur`, `l`.`Pref_Tel_Dom`, `l`.`Prenom`, `l`.`Ref_Locataire`, `l`.`Suite_Adresse`, `l`.`Tel_Bur`, `l`.`Tel_Dom`, `l`.`Ville`, ROW_NUMBER() OVER(PARTITION BY `l`.`RefAgence`, `l`.`No_Locataire` ORDER BY `l`.`RefAgence`, `l`.`No_Locataire`) AS `row`
              FROM `locat` AS `l`
          ) AS `t`
          WHERE `t`.`row` <= 1
      ) AS `t0` ON (`r`.`RefAgence` = `t0`.`RefAgence`) AND (`r`.`No_Locataire` = `t0`.`No_Locataire`)
fail: Microsoft.EntityFrameworkCore.Query[10100]
      An exception occurred while iterating over the results of a query for context type 'Eos.Data.EFCore.ACDVReadContext'.
      MySql.Data.MySqlClient.MySqlException (0x80004005): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(PARTITION BY `l`.`RefAgence`, `l`.`No_Locataire` ORDER BY `l`.`RefAgence`, `l`.' at line 32
       ---> MySql.Data.MySqlClient.MySqlException (0x80004005): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(PARTITION BY `l`.`RefAgence`, `l`.`No_Locataire` ORDER BY `l`.`RefAgence`, `l`.' at line 32
         at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ResultSet.cs:line 49
         at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 125
         at MySql.Data.MySqlClient.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 338
         at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\CommandExecutor.cs:line 63
         at MySql.Data.MySqlClient.MySqlCommand.ExecuteDbDataReader(CommandBehavior behavior) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 218
         at System.Data.Common.DbCommand.ExecuteReader()
         at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReader(RelationalCommandParameterObject parameterObject)
         at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
MySql.Data.MySqlClient.MySqlException (0x80004005): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(PARTITION BY `l`.`RefAgence`, `l`.`No_Locataire` ORDER BY `l`.`RefAgence`, `l`.' at line 32
 ---> MySql.Data.MySqlClient.MySqlException (0x80004005): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(PARTITION BY `l`.`RefAgence`, `l`.`No_Locataire` ORDER BY `l`.`RefAgence`, `l`.' at line 32
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ResultSet.cs:line 49
   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 125
   at MySql.Data.MySqlClient.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 338
   at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\CommandExecutor.cs:line 63
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteDbDataReader(CommandBehavior behavior) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 218
   at System.Data.Common.DbCommand.ExecuteReader()
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReader(RelationalCommandParameterObject parameterObject)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
MySql.Data.MySqlClient.MySqlException (0x80004005): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(PARTITION BY `l`.`RefAgence`, `l`.`No_Locataire` ORDER BY `l`.`RefAgence`, `l`.' at line 32
 ---> MySql.Data.MySqlClient.MySqlException (0x80004005): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(PARTITION BY `l`.`RefAgence`, `l`.`No_Locataire` ORDER BY `l`.`RefAgence`, `l`.' at line 32
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ResultSet.cs:line 49
   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 125
   at MySql.Data.MySqlClient.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 338
   at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\CommandExecutor.cs:line 63
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteDbDataReader(CommandBehavior behavior) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 218
   at System.Data.Common.DbCommand.ExecuteReader()
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReader(RelationalCommandParameterObject parameterObject)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Eos.Api.Helpers.G2AHelper.PrecomputeG2AOccupations(String agencyId, ACDVReadContext readContext, ACDVWriteContext writeContext, DateTime start, DateTime end, CancellationToken cancellationToken) in /home/fmauneko/code/eos/Eos.Api/Helpers/G2AHelper.cs:line 112
   at Eos.Api.Controllers.Channels.G2AController.GetG2AOccupations(String agencyId, DateTime startDate, DateTime endDate, CancellationToken cancellationToken) in /home/fmauneko/code/eos/Eos.Api/Controllers/Channels/G2AController.cs:line 75
   at Eos.Api.Controllers.Channels.G2AController.GetG2AOccupations(String agencyId, CancellationToken cancellationToken) in /home/fmauneko/code/eos/Eos.Api/Controllers/Channels/G2AController.cs:line 58
   at lambda_method(Closure , Object )
   at Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable.Awaiter.GetResult()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.<Invoke>g__AwaitMatcher|8_0(EndpointRoutingMiddleware middleware, HttpContext httpContext, Task`1 matcherTask)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

### Further technical details

MySQL version: 5.7.27
Operating system: Linux x64 
Pomelo.EntityFrameworkCore.MySql version: 3.0.0-rc1-final
Microsoft.AspNetCore.App version: 3.0.0

**Comment by lauxjpn** [comment]
Without the corresponding model definitions and a SQL dump of the related tables and some sample data, I cannot make any comments on *why* EF Core 3.0 is generating a `ROW_NUMBER() OVER(PARTITION BY)` clause.

The fact that is does however, is not compatible with MySQL < 8.0.0 as you correctly state.

At the moment, your only option is to rewrite your query so it will be translated into SQL without making use of window functions.

It would be preferable though, if we would actually intercept this behavior during the query pipeline before it leads to a MySQL syntax error, so we could properly let EF Core either throw the client eval exception (if client eval is disabled) or let EF Core fallback on client eval (if client eval is enabled).

We keep this issue open to investigate, whether a better client eval support can be implemented.

**Comment by ApocDev** [comment]
Why is this marked as closed? This is still happening, and "change your query" is not a suitable workaround.

```
            var first = await Db.Sessions
                .Where(s => s.Id == session.Value)
                .Select(s => s.Accounts.AssociatedIds.FirstOrDefault())
                .FirstOrDefaultAsync();
```

Generates the following query:

```
// @__session_Value_0='08d75bbe-aea6-2466-7fae-785e6a87fd87'

SELECT `t0`.`Id`, `t0`.`AccountId`, `t0`.`AccountLookupId`, `t0`.`CustomerId`, `t0`.`LU`
FROM `Sessions` AS `s`
LEFT JOIN `AccountLookups` AS `a` ON `s`.`AccountsId` = `a`.`Id`
LEFT JOIN (
    SELECT `t`.`Id`, `t`.`AccountId`, `t`.`AccountLookupId`, `t`.`CustomerId`, `t`.`LU`
    FROM (
        SELECT `b`.`Id`, `b`.`AccountId`, `b`.`AccountLookupId`, `b`.`CustomerId`, `b`.`LU`, ROW_NUMBER() OVER(PARTITION BY `b`.`AccountLookupId` ORDER BY `b`.`Id`) AS `row`
        FROM `AssociatedIds` AS `b`
    ) AS `t`
    WHERE `t`.`row` <= 1
) AS `t0` ON `a`.`Id` = `t0`.`AccountLookupId`
WHERE (`s`.`Id` = @__session_Value_0) AND @__session_Value_0 IS NOT NULL
LIMIT 1
```

Setting the server version should be disabling the OVER(PARTITION ..) entirely for anything < 8.0.

**Comment by lauxjpn** [comment]
@ApocDev 

This was fixed with #871 (after releasing `3.0.0-rc1.final`) and will be published on NuGet with `3.0.0-rc2.final`. It is currently only part of the nightly builds.

What version of Pomelo.EntityFrameworkCore.MySql are you using?

Fixed in this case means, that it will throw the EF Core client eval exception (translation error exception) for MySQL < 8.0, instead of the syntax error exception that is thrown in Pomelo < `3.0.0-rc1.final`.

This does not change the fact though, that you have to "change your query", to avoid an exception, as MySQL < 8.0 does not support window functions. The only difference to Pomelo <= `2.2.6` is, that `2.2.6` was used in conjunction with EF Core 2.2.x which used a different (less efficient) query pipeline under the hood.

To change your query, either just simplify it, so that EF Core's query pipeline does not need to use window functions, or trigger [client evaluation](https://docs.microsoft.com/en-us/ef/core/querying/client-eval) at some point in the query, to let LINQ to objects do the work unsupported by your MySQL version.
